src/os/bluestore/BlueStore.cc

```diff
@@ -1170,41 +1173,42 @@ struct LruOnodeCacheShard : public BlueStore::OnodeCacheShard {
     ceph_assert(p != lru.begin());
     --p;
     ceph_assert(num >= n);
-    num -= n;
     while (n-- > 0) {
       BlueStore::Onode *o = &*p;
       dout(20) << __func__ << "  rm " << o->oid << " "
-               << o->nref << " " << o->cached << " " << o->pinned << dendl;
+               << o->nref << " " << o->cached << dendl;
+
       if (p != lru.begin()) {
         lru.erase(p--);
       } else {
         ceph_assert(n == 0);
         lru.erase(p);
       }
       *(o->cache_age_bin) -= 1;
-      auto pinned = !o->pop_cache();
-      ceph_assert(!pinned);
-      o->c->onode_space._remove(o->oid);
+      if (o->pin_nref > 1) {
```

Comment:
Could you elaborate a bit on what could cause the pin_nref > 1 here?

Reply:
pin_nref is incremented at Onode::get along with regular onode's ref counter. So we might have pin_nref > 1 before we get into this trim_to method and the latter reaches the state where it might want to trim this onode. 
So you can just treat pin_nref as almost a full replica of nref - the only divergence actually occurs at Onode::put: pin_nref is decremented at the entrance of this func and nref one at the end.

As for map removal - being in the map means onode is in the cache - so we can remove from the onode when onode is to be destroyed or removed from the cache. Pinning doesn't mean cache removal but forces removal from cache LRU to avoid repetitive (and useless) scans over pinned entries during the following trim attempts - we saw such scans might dramatically impact operations latencies before.
Hence if pin_nref >1 (aka onode is pinned) the trimming is not allowed to delete the onode completely but exclusively from LRU.
While if pin_nref is at 1 - it should be removed from both LRU and cache map.

Reply:
Right,  my confusion was that shouldn't we do the pinning logic inside this if statement?  And I couldn't find anyplace else you do the pinning logic, wouldn't this make the Onode stay in the LRU even if it's logically pinned?

Reply:
Ah, I guess you meant to call do the pin here but missed it?

Reply:
Nope. An onode is unconditionally removed from LRU list a couple lines above - if we reach an onode at this point - it should be either trimmed out or pinned. Hence LRU removal happens anyway. But when pin_nref >1 the onode to be  logically preserved in the cache (aka onode_space) - hence we do nothing and onode becomes pinned (i.e. present in the cache but not observed during trimming due to its non-existence in LRU). Contrary full trimming does removal from onode_space as well and this (implicitly, due to ref counting) releases onode completely.

Reply:
Gotcha, thanks for your detailed explanation